Tutorial: How to create custom surveys (#6373)
PostHog/posthog.com: contents/tutorials/survey.md

![Survey in-app](../images/tutorials/survey/survey.png)

## Creating a custom survey

If you want to customize how and when you show your survey, we can can that too. To start, create another survey, but this time set the "Display mode" as `API` instead of `popover`. Customize the details as much as you want, save it as a draft, then launch. It won’t show in your app until we write some code to implement it.

![API display mode](../images/tutorials/survey/api.png)

Once done, head back to the app and remove the boilerplate from `page.js` . We’ll make this the page with our survey. To start, this requires the `use client` directive, `usePostHog` context, and a `useState` hook for the survey data.

```js-web
// app/page.js
'use client'
import { usePostHog } from 'posthog-js/react'
import { useState } from 'react'

export default function Home() {
  
  const posthog = usePostHog()
  const [survey, setSurvey] = useState(null)

  return (
    <div>
      <h1>Take our survey!</h1>
    </div>
  )
}
```

> We can then get the surveys for a user with PostHog’s `getActiveMatchingSurveys` method in a `useEffect`. In the callback, filter for API type surveys, set the first one in our state, and capture a `survey shown` event.

Reviewer: ```suggestion
We can then get the surveys for a user with PostHog’s `getActiveMatchingSurveys` method in a `useEffect`. In the callback, filter for API type surveys, set the first one in our state, and capture a `survey shown` event:
```

      <div>
        <h2>{survey.name}</h2>
        <h4>{survey.questions[0].question}</h4>
        <textarea
          value={textAreaValue}
          onChange={handleTextAreaChange}
        ></textarea>
//...
```

Finally, set up a `survey sent` event capture when the user clicks the submit button. This needs to include the survey ID, name, and response, and should also reset the `textarea`.

```js-web
// app/page.js
//...
	const submit = () => {
    posthog.capture("survey sent", {
      $survey_id: survey.id,
      $survey_name: survey.name,
      $survey_response: textAreaValue
    })
    setTextAreaValue('')
  }

  return (
    <div>
      <h1>Take our survey!</h1>
      {survey && (
        <div>
          <h2>{survey.name}</h2>
          <h4>{survey.questions[0].question}</h4>
          <textarea
            value={textAreaValue}
            onChange={handleTextAreaChange}
          ></textarea>
          <br/>
          <button onClick={submit}>Submit</button>
        </div>
      )}
    </div>
  )
}
```

When you submit the survey with a response, it shows up in your results in PostHog.

![Results](../images/tutorials/survey/results.png)

### Dismiss and hide survey

> The last thing you might want to do is let users  dismiss the survey and prevent it from being shown repeatedly. The PostHog survey app handles this with `localStorage` storing a value that is `seenSurvey_ + survey.id`, so we will use that as well.

Reviewer: ```suggestion
The last thing you might want to do is enable users to dismiss the survey and prevent it from being shown repeatedly. The PostHog survey app handles this with `localStorage` storing a value that is `'seenSurvey_' + survey.id`, so we will use that as well.
```

First, create a function to close the survey which sets the value in `localStorage` and the state of the survey to `null`.

```js-web
// app/page.js
//...
const closeSurvey = () => {
  localStorage.setItem('seenSurvey_' + survey.id, true)
  setSurvey(null)
}
//...
```

Next, add the `closeSurvey` function to the `submit` function and a new `dismiss` function that captures a `survey dismissed` event. We also need a dismiss button on our page under our submit button.

```js-web
// app/page.js
//...

const submit = () => {
  posthog.capture("survey sent", {
    $survey_id: survey.id,
    $survey_name: survey.name,
    $survey_response: textAreaValue
  })
  setTextAreaValue('')
  closeSurvey()
}

const dismiss = () => {
  posthog.capture("survey dismissed", {
    $survey_id: survey.id,
    $survey_name: survey.name,
  })
  closeSurvey()
}

//...
<button onClick={submit}>Submit</button>
<br/>
<button onClick={dismiss}>Dismiss</button>
//...
```

> This hides the survey when either are clicked, but it shows again when the page is refreshed. To fix this, wecheck for the `seenSurvey_` value in local storage and don’t show the survey if it’s `true`. We can do this in the `useEffect` like this:

Reviewer: ```suggestion
This hides the survey when either are clicked, but it shows again when the page is refreshed. To fix this, we check for the `seenSurvey_` value in local storage and don’t show the survey if it’s `true`. We can do this in the `useEffect` like this:
```
